### ISSUE TITLE
Node undefined

### ISSUE THREAD

**Issue by still-dreaming-1** [metakirby5/codi.vim#17]
When I use this to run JavaScript code in node, the scratch buffer is showing `undefined` in strange places. When I run the same code in node, I don't get any warnings or errors.

**Comment by still-dreaming-1** [comment]
I just noticed when I leave a buffer and come back, the `undefined` messages move around. Sometimes they disappear. Sometimes they are on different lines than they were before.

**Comment by iovis9** [comment]
A way to reproduce it would help.

**Comment by metakirby5** [comment]
Agreed - if you can provide the following, I'll definitely be able to fix it much faster:

- The entire process (from opening Vim to producing the issue)
- What you put into the buffer
- A screencast, if possible
- A `.vimrc` stripped down to the bare minimum to reproduce the issue, if possible

**Comment by StephenGilboy** [comment]
So the way to reproduce it is to use `var`.
![screenshot 2016-08-16 12 11 08](https://cloud.githubusercontent.com/assets/827735/17712410/a01af32e-63aa-11e6-9ffb-416f328ac67a.png)
![screenshot 2016-08-16 12 11 35](https://cloud.githubusercontent.com/assets/827735/17712411/a2597e80-63aa-11e6-8d3a-264ee03bb278.png)

**Comment by metakirby5** [comment]
If you try using `var` in node, you get undefined as well. The intended behavior is to match what the underlying interpreter would produce, so this is expected and will not be fixed.

The undefined messages moving around is a bit concerning though - can you post an example of that?

**Comment by still-dreaming-1** [comment]
Ah, yes your right. I tried requiring the file from node, not manually typing in the code from the file. When I do that it does say things like undefined in those places. It just means nothing was returned, it is not an error or warning, just the repl printing what was returned. It seems strange that the REPL expects anything to be returned in when you defining variables, but that has nothing to do with codi.

**Comment by obxhdx** [comment]
@still-dreaming-1

Node provides an option called `ignoreUndefined` ([source](https://nodejs.org/api/repl.html#repl_repl_start_options)) that you could use to work around this. Here's what I did:

First create a custom REPL that sets `ignoreUndefined` to true:

```javascript
#!/usr/bin/env node

var repl = require('repl'),
    replOpts = { prompt: '> ', ignoreUndefined: true },
    replServer = repl.start(replOpts);
```

Save it in a file somewhere with execution rights:
```
$ chmod +x path/to/my-repl.js
```

Then configure Codi to use your custom REPL instead `node`:

```vim
let g:codi#interpreters = {
      \  'javascript': {
      \     'bin': 'path/to/my-repl.js',
      \     'prompt': '^\(>\|\.\.\.\+\) ',
      \   },
      \ }
```

**Comment by still-dreaming-1** [comment]
@obxhdx That seems to be the best solution if we don't want to see undefined, which is saying something bad about the node REPL. I actually don't want to completely disable showing undefined in the REPL when that is what is returned, but I don't want node to do ridiculous things like show undefined when defining a variable. It should output the contents of the variable like any normal REPL would. Anyway, that has nothing to do with codi. Thanks for the tip.

**Comment by metakirby5** [comment]
FYI, this was resolved with #42.
